### PR TITLE
Fix gradient accumulation in post training sft

### PR DESCRIPTION
**Pull request by ChingTsai**
# Description

FIXES: b/478823561

This PR resolves discrepancies in loss calculation and training step counts when running SFT with gradient accumulation enabled. Currently, `MaxText.sft.sft_trainer` (which uses the Tunix trainer) exhibits breaking behavior when gradient accumulation is turned on, diverging significantly from the native implementation in `MaxText.sft_trainer`. This change aligns the Tunix-based SFT logic to match the native behavior.

## Problem Statement

- Loss Disparity: When GA is enabled, MaxText-Tunix observed a massive loss scale disparity compared to the native implementation. This occurs because MaxText-Tunix uses the same loss_fn as the native implementation. The native logic explicitly skips dividing by total_weights inside the function (deferring normalization to a later stage [here](https://github.com/AI-Hypercomputer/maxtext/pull/3040/changes#diff-4956f1016541609a6ae02b50cc212b3ba30cffc67e40d929e85dffaa0ce82baeR174-R178)). Consequently, Tunix inherited this behavior and failed to normalize the loss, resulting in broken calculations and inflated values. 

pr_fix_vs_original_vs_native

- Step Count Mismatch: While MaxText-Native handles micro-batching internally by reshaping the full global batch, Tunix relies on the input pipeline to provide pre-sized micro-batches. Without this adjustment, Tunix was ingesting full global batches at every step, resulting in incorrect epoch calculations and causing the run to terminate prematurely compared to the native implementation.

## Changes
- Introduced a new configuration `use_tunix`
  -  Determine whether gradient accumulation (GA) is performed through the Tunix implementation.

- Modified the loss function
  - Aligned Tunix loss calculation with native behavior. these changes are backward-compatible and will not affect existing native GA implementations.

- Pre-sliced batches in data preprocessing
  - Ensures that when using Tunix GA, the chunk size is correctly calculated during the data preparation stage.

- Add Tunix SFT integration test for gradient accumulation loss verification [here](https://github.com/AI-Hypercomputer/maxtext/pull/3040/changes#diff-7173b581a58bb895b89d15af0c2474889d2f969215b014053b0d5f2bd475019fR159-R224)

# Tests
## End2End
```
python3 -m MaxText.sft.sft_trainer \
    src/MaxText/configs/sft.yml \
    run_name=$RUN_NAME \
    base_output_directory=..../qwen3-4b \
    model_name=qwen3-4b \
    load_parameters_path=..../qwen3-4b/0/items \
    tokenizer_path=Qwen/qwen3-4b \
    steps=$train_step \
    profiler=xplane \
    hf_path=arrow \
    dataset_type=hf \
    train_split=train \
    hf_train_files=..../data-00000-of-00001.arrow \
    hf_eval_files=..../data-00000-of-00001.arrow \
    per_device_batch_size=4 \
    gradient_accumulation_steps=4 \
    max_target_length=1024 \
    learning_rate=1.3e-5 \
    warmup_steps_fraction=0.05 \
    data_shuffle_seed=42 \
    gradient_clipping_threshold=1 \
    learning_rate_final_fraction=0 \
    weight_dtype=bfloat16
```
After applying the changes, the loss graphs of both versions are now almost identical.

graph_2026-01-29_15-52-15

## Integration Test
```
python -m unittest tests.integration.gradient_accumulation_test.GradientAccumulationTest.test_tunix_sft_grad_accumulate_same_loss
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/trainers/post_train/sft/train_sft.py [hunk 1]
```diff
@@ -194,7 +194,7 @@ def main(argv: Sequence[str]) -> None:
   pathwaysutils.initialize()
   os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
 
-  mt_config = pyconfig.initialize(argv)
+  mt_config = pyconfig.initialize(argv, use_tunix_ga=True)
```
Comment (SurbhiJainUSC) [inline src/maxtext/trainers/post_train/sft/train_sft.py]: Can `use_tunix_ga` be part of SFT configs: https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/configs/sft.yml? It can be set to True by default. Also, we can rename it to `use_tunix`.

File: src/MaxText/sft_trainer.py [hunk 2]
```diff
@@ -165,7 +165,7 @@ def main(argv: Sequence[str]) -> None:
     os.environ["LIBTPU_INIT_ARGS"] = (
         os.environ.get("LIBTPU_INIT_ARGS", "") + " --xla_tpu_spmd_rng_bit_generator_unsafe=true"
     )
-  config = pyconfig.initialize(argv)
+  config = pyconfig.initialize(argv, use_tunix=False)
```
Comment (ChingTsai) [inline src/MaxText/sft_trainer.py]: Add an override here to maintain backward compatibility with the native `sft_trainer`.

File: src/maxtext/trainers/post_train/sft/train_sft.py (hunk 1 above)
Comment (ChingTsai) [reply src/maxtext/trainers/post_train/sft/train_sft.py]: Fixed.

File: src/MaxText/configs/types.py [hunk 3]
```diff
@@ -1055,6 +1055,10 @@ class Optimizer(BaseModel):
   gradient_accumulation_steps: PositiveInt = Field(
       1, description="Number of steps to accumulate gradients before updating."
   )
+  use_tunix: bool = Field(
+      False,
+      description="Whether to use the Tunix implementation for gradient accumulation.",
+  )
```
Comment (NuojCheng) [inline src/MaxText/configs/types.py:1062]: I think the name `use_tunix` is too broad. If we are only interested in cases for sft + tunix + GA, please use a more specific name, e.g. `use_tunix_gradient_accumulation`.

File: src/MaxText/input_pipeline/_hf_data_processing.py [hunk 4]
```diff
@@ -54,7 +54,9 @@ def vision_sft_preprocessing_pipeline(
   """pipeline for multimodal SFT with HF dataset"""
 
   assert len(text_columns) == 2, f"Need two text_columns for query and response, received {text_columns=}"
-  batch_size = global_batch_size // jax.process_count()
+  # Tunix GA requires per-micro-batch slicing at the data level,
+  # whereas Native GA processes the full batch and splits it internally.
+  batch_size = global_batch_size // jax.process_count() // (config.gradient_accumulation_steps if config.use_tunix else 1)
```
Comment (NuojCheng) [inline src/MaxText/input_pipeline/_hf_data_processing.py]: nit: I would perfer
```
if config.use_tunix:
   batch_size = ()
else:
  batch_size = ()
```

File: src/MaxText/input_pipeline/_hf_data_processing.py [hunk 5]
```diff
@@ -195,13 +197,18 @@ def preprocessing_pipeline(
...
     sft_train_on_completion_only=True,
     grain_worker_count=1,  # only support 0 or 1
     max_segments_per_seq=None,
 ):
   """pipeline for preprocessing HF dataset"""
 
   assert global_batch_size % global_mesh.size == 0, "Batch size should be divisible by number of global devices."
+  # Tunix GA requires per-micro-batch slicing at the data level,
+  # whereas Native GA processes the full batch and splits it internally.
+  batch_size = global_batch_size // jax.process_count() // (num_microbatches if use_tunix else 1)
```
Comment (NuojCheng) [inline src/MaxText/input_pipeline/_hf_data_processing.py]: same with comment above

File: src/MaxText/train.py [hunk 6]
```diff
@@ -179,7 +179,7 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
   # Zero1+GA to reduce communication overhead.
   # EPS was used to avoid division by zero, but it's not needed when gradient
   # accumulation is enabled since there's no division.
-  if config.gradient_accumulation_steps > 1:
+  if config.gradient_accumulation_steps > 1 and not config.use_tunix:
```
Comment (NuojCheng) [inline src/MaxText/train.py]: please add comments explaining why we don't follow standard GA path when `use_tunix=true`

**Review by NuojCheng — COMMENTED** [review]
It is a bit concerning to me why this PR fails tests/integration/train_tests.py::TrainTests::test_tpu_zero1_gradient_accumulation with OOM. The changes made by this PR should not affect regular GA performance in `train.py`. We will need to take a look on xprofs to figure out why.

File: src/MaxText/input_pipeline/_hf_data_processing.py (hunk 4 above)
Comment (richjames0) [reply src/MaxText/input_pipeline/_hf_data_processing.py]: Agree. I love expressions over statements but this is a bit complex for the former

**Comment by ChingTsai** [comment]
> It is a bit concerning to me why this PR fails tests/integration/train_tests.py::TrainTests::test_tpu_zero1_gradient_accumulation with OOM. The changes made by this PR should not affect regular GA performance in `train.py`. We will need to take a look on xprofs to figure out why.

Hi @NuojCheng and @richjames0 ,

Thanks for the review! I have addressed your comments in the latest push.

Regarding the failing `tests/integration/train_tests.py::TrainTests::test_tpu_zero1_gradient_accumulation` OOM test. Is it possible the shared runner is competing for resources with other jobs? Or could some JAX arrays be leaking (not being garbage collected) during continuous testing? I’ve seen this test pass intermittently, for example in [here](https://github.com/AI-Hypercomputer/maxtext/actions/runs/21623645517/job/62318366296#step:7:57).
I'm happy to investigate further if you think it's related to the current changes.

**Comment by ChingTsai** [comment]
hi @NuojCheng, could you help to merge this PR? Thanks!
cc @chishuen

**Comment by ChingTsai** [comment]
Hi @NuojCheng ,

Although all integration tests are passed, I noticed that the TPU integration test is now taking significantly longer (> 70 min).
After investigating, It seems like the Tunix test might be leaving the testing process in a state that slows down subsequent tests. If I move that specific test to the end of the suite, the runtime returns to normal, as seen in [this run](https://github.com/AI-Hypercomputer/maxtext/actions/runs/21663656103/job/62453809768#step:7:90).

Should I pull this test from this PR and reintroduce it in a separate PR once the root cause is fixed?

**Comment by NuojCheng** [comment]
> Hi @NuojCheng ,
> 
> Although all integration tests are passed, I noticed that the TPU integration test is now taking significantly longer (> 70 min). After investigating, It seems like the Tunix test might be leaving the testing process in a state that slows down subsequent tests. If I move that specific test to the end of the suite, the runtime returns to normal, as seen in [this run](https://github.com/AI-Hypercomputer/maxtext/actions/runs/21663656103/job/62453809768#step:7:90).
> 
> Should I pull this test from this PR and reintroduce it in a separate PR once the root cause is fixed?

Thank you for the investigation! Please create a separate PR.

I noticed the [pylint test fails](https://github.com/AI-Hypercomputer/maxtext/actions/runs/21631296934/job/62344455697?pr=3040). Please use `pre-commit` to fix it. Check `.pre-commit-config.yaml` for details.
